qiskit/algorithms/evolvers/classical_methods/scipy_real_evolver.py

```diff
@@ -0,0 +1,328 @@
+        else:
+            state = evolution_problem.initial_state.to_matrix(massive=True).transpose()
+
+        hamiltonian = evolution_problem.hamiltonian.to_spmatrix()
+
+        # Determine the number of timesteps.
+        timesteps = min(
+            self._ntimesteps(
+                time=evolution_problem.time, hamiltonian=hamiltonian, threshold=self.threshold
+            ),
+            self.max_iterations,
+        )
+        timestep = evolution_problem.time / timesteps
+
+        # Create the operators for the time evolution.
+        idnty = sp.identity(hamiltonian.shape[0], format="csr")
+
+        lhs_operator = idnty + 1j * timestep / 2 * hamiltonian
+        rhs_operator = idnty - 1j * timestep / 2 * hamiltonian
+
+        if isinstance(evolution_problem.aux_operators, list):
+            aux_ops = [aux_op.to_spmatrix() for aux_op in evolution_problem.aux_operators]
+        elif isinstance(evolution_problem.aux_operators, dict):
+            aux_ops = [aux_op.to_spmatrix() for aux_op in evolution_problem.aux_operators.values()]
+        else:
+            aux_ops = []
+
+        aux_ops_2 = [aux_op.dot(aux_op) for aux_op in aux_ops]
+
+        return (
+            state,
+            lhs_operator,
+            rhs_operator,
+            aux_ops,
+            aux_ops_2,
+            timesteps,
+            self._bicg_tol(timesteps),
+        )
+
+    def _bicg_tol(self, timesteps: int) -> float:
+        """Returns the tolerance for the BiCG solver.
+
+        Args:
+            timesteps: The number of timesteps in the evolution.
+
+        Returns:
+            The tolerance for the BiCG solver.
+        """
+        return self.threshold * self.bicg_err / timesteps
+
+    def minimal_number_steps(
+        self, norm_hamiltonian: float, time: float, threshold: float = 1e-4
+    ) -> int:
+        r"""Calculate the timestep for the given time and threshold.
+
+        We use the fact that the Taylor expansion term of third order in our expansion would be
+        :math:`\frac{ ( -i H \delta t ) ^3}{12}` to compute an approximation for how
+        many timesteps we need to reach a certain precision.
+
+        Args:
+            time: The time to evolve.
+            norm_hamiltonian: The operator norm of the Hamiltonian if known or an estimation of
+                             it (For example the infinity norm of the Hamiltonian). This value
+                             will be associated with the error of the Hamiltonian.
+            threshold: The threshold for the error.
+
+        Returns:
+            The number of timesteps needed to reach the error threshold.
+        """
+        return int(
+            np.power(norm_hamiltonian * time, 3 / 2)
+            * np.power(12 * threshold * (1 - self.bicg_err), -1 / 2)
+            + 1
+        )
+
+    def _ntimesteps(self, time: float, hamiltonian: sp.csr_matrix, threshold: float = 1e-4) -> int:
```

Comment:
Not sure if we need this, since we already have `minimal_number_steps` -- could we use `minimal_number_steps` directly instead of this private method?

Reply:
So, minimal_number_steps is a function that will estimate the ammount of timesteps needed to reach certain precision given the norm of the hamiltonian. Ideally we would want to use the highest absolute value eigenvalue in order to compute this, since this is the tightest bound on the error associated to mutiplying by H. However, that would be too expensive so instead, in _ntimesteps() we use the infinity norm which is a much looset bound but very easy to compute.
Therefore if the user has previous knowledge of the hamiltonian and they can provide an estimation of the highest eigenvalue, they are better off just providing that to minimal_number_steps() manually before setting the maximum_number_steps.

Therefore I think it makes sense to keep both.